### PR TITLE
Add Linux support for illuminance sensors

### PR DESCRIPTION
**Pull request by JesseTG**
## Description

This pull requests implements support for illuminance sensors on Linux; the reading is exposed through `retro_sensor_interface` and `RETRO_SENSOR_ILLUMINANCE`. I used my Steam Deck to develop and test it, but this should work on any Linux device that has a light sensor.

Here's a video of it in action:

https://www.youtube.com/watch?v=0cCWS5X3Jcg

Some details about the implementation:

- This integration is available for all Linux-supported input drivers (including `sdl2`) except `wayland`. The implementation is orthogonal to the other controller drivers, as illuminance sensors [use a different interface](https://github.com/torvalds/linux/blob/master/Documentation/ABI/testing/sysfs-bus-iio#L1616-L1627) than gamepads.
- The solar sensor is polled on a separate thread, with the rate set by the `rate` parameter of `retro_set_sensor_state_t`. If the rate is zero, it defaults to 5 times a second. I don't use locks, as I'm assuming that aligned 4-byte reads and writes are atomic on the CPUs that would run this patch. **If this assumption is incorrect, please let me know.**
- This integration is primarily intended for melonDS DS and mGBA, as the games in the Boktai franchise (including [Lunar Knights](https://en.wikipedia.org/wiki/Lunar_Knights)) used a built-in light sensor to grant buffs to the player. mGBA already supports `retro_set_sensor_state_t`, and I'm working on it for melonDS DS.

## Related Pull Requests

https://github.com/libretro/RetroArch/pull/16188 touches the same interface (though not for illuminance), and will likely have to be updated to address merge conflicts.

### CODE REVIEW

**Comment by JesseTG** [comment]
Oh, one more thing; if the sensor is closed while the poll thread is in `retro_sleep`, the main thread will be blocked until the poll thread wakes up. Let me do something about that.

**Comment by JesseTG** [comment]
I think I'll use `pthread_cancel` and `pthread_setcancelstate` to allow the sensor polling thread to be canceled if (and _only_ if) it's in the middle of `retro_sleep`; POSIX requires `nanosleep` (which `retro_sleep` uses) to be a cancellation point, so it should be okay.

**Comment by JesseTG** [comment]
> If you want to do any threading at all I'd have to suggest using rthreads instead, at least that way we are not tied/dependent on pthread.

Do we support any Linux environments that _don't_ provide pthread? Remember, the functionality in this PR is fundamentally Linux-specific.

> Anyway is this PR good to merge?

Almost. First I want to solve the problem of blocking the thread when closing the sensor.

**Comment by JesseTG** [comment]
@LibretroAdmin All right, I'm good to go. Thanks for checking this out.

**Comment by LibretroAdmin** [comment]
Alright I'll give it a look over the next day or so.

File: input/common/linux_common.c [hunk 1]
```diff
@@ -14,22 +14,57 @@
...
+   /* We store the lux reading in millilux (as an int)
+    * so that we can make the access atomic and avoid locks.
+    * A little bit of precision is lost, but not enough to matter.
+    */
+   volatile int millilux;
+
+   /* If true, the associated thread must finish its work and exit. */
+   volatile bool done;
+
+   char path[PATH_MAX];
```
Comment (LibretroAdmin) [inline input/common/linux_common.c:64]: Use here either PATH_MAX_LENGTH, NAME_MAX_LENGTH or DIR_MAX_LENGTH, depending on your size needs.

File: input/common/linux_common.c [hunk 2]
```diff
@@ -121,3 +156,222 @@ bool linux_terminal_disable_input(void)
...
+
+   sensor->millilux  = 0;
+   sensor->poll_rate = rate ? rate : DEFAULT_POLL_RATE;
+   sensor->thread    = NULL; /* We'll spawn a thread later, once we find a sensor */
+   sensor->done      = false;
+
+   devices = opendir(IIO_DEVICES_DIR);
+   if (!devices)
+   { /* If we couldn't find the IIO device directory... */
+      char errmesg[PATH_MAX];
```
Comment (LibretroAdmin) [inline input/common/linux_common.c:213]: Same here, see previous comment

File: input/common/linux_common.c [hunk 3]
```diff
@@ -121,3 +156,222 @@ bool linux_terminal_disable_input(void)
...
+      int err = errno;
+      double lux;
+
+      if (d->d_name[0] == '.')
+         /* Skip hidden files, ".", and ".." */
+         continue;
+
+      if (err != 0)
+      {
+         char errmesg[PATH_MAX];
```
Comment (LibretroAdmin) [inline input/common/linux_common.c:239]: Same here.

File: input/common/linux_common.c [hunk 4]
```diff
@@ -121,3 +156,222 @@ bool linux_terminal_disable_input(void)
...
+
+   if (sensor->thread)
+   {
+      pthread_t thread = sthread_get_thread_id(sensor->thread);
+      sensor->done = true;
+
+      if (pthread_cancel(thread) != 0)
+      {
+         int err = errno;
+         char errmesg[PATH_MAX];
```
Comment (LibretroAdmin) [inline input/common/linux_common.c:294]: Same here

File: input/common/linux_common.c [hunk 5]
```diff
@@ -121,3 +156,222 @@ bool linux_terminal_disable_input(void)
...
+   int err;
+   FILE* in_illuminance_input = NULL;
+
+   if (!sensor || sensor->path[0] == '\0')
+      return -1.0;
+
+   in_illuminance_input = fopen(sensor->path, "r");
+   if (!in_illuminance_input)
+   {
+      char errmesg[PATH_MAX];
```
Comment (LibretroAdmin) [inline input/common/linux_common.c:343]: Same here.

File: input/common/linux_common.c [hunk 6]
```diff
@@ -121,3 +156,222 @@ bool linux_terminal_disable_input(void)
...
+      goto done;
+   }
+
+   /* TODO: This may be locale-sensitive */
+   errno = 0;
+   illuminance = strtod(buffer, NULL);
+   err = errno;
+   if (err != 0)
+   {
+      char errmesg[PATH_MAX];
```
Comment (LibretroAdmin) [inline input/common/linux_common.c:365]: Same here.

File: input/drivers/udev_input.c [hunk 7]
```diff
@@ -4013,9 +4015,59 @@ static void udev_input_free(void *data)
...
+         if (udev->illuminance_sensor)
+            /* If we already have a sensor, just set the rate */
+            linux_set_illuminance_sensor_rate(udev->illuminance_sensor, rate);
+         else
+            udev->illuminance_sensor = linux_open_illuminance_sensor(rate);
+
+         return udev->illuminance_sensor != NULL;
+      default:
+         return false;
+   }
```
Comment (LibretroAdmin) [inline input/drivers/udev_input.c:4051]: This function should ALWAYS return something near the end. Make the 'default' return false; the last line here and just do a 'break;' at the default switch case instead.

File: input/drivers/udev_input.c [hunk 8]
```diff
@@ -4013,9 +4015,59 @@ static void udev_input_free(void *data)
...
+      return 0.0f;
+
+   switch (id)
+   {
+      case RETRO_SENSOR_ILLUMINANCE:
+         if (udev->illuminance_sensor)
+            return linux_get_illuminance_reading(udev->illuminance_sensor);
+      default:
+         return 0.0f;
+   }
```
Comment (LibretroAdmin) [inline input/drivers/udev_input.c:4068]: Same here.

File: input/drivers/x11_input.c [hunk 9]
```diff
@@ -431,7 +436,66 @@ static void x_input_free(void *data)
...
+         if (x11->illuminance_sensor)
+           /* If we already have a sensor, just set the rate */
+           linux_set_illuminance_sensor_rate(x11->illuminance_sensor, rate);
+         else
+           x11->illuminance_sensor = linux_open_illuminance_sensor(rate);
+
+         return x11->illuminance_sensor != NULL;
+#endif
+      default:
+         return false;
```
Comment (LibretroAdmin) [inline input/drivers/x11_input.c:478]: Same here.

File: input/drivers/x11_input.c [hunk 10]
```diff
@@ -431,7 +436,66 @@ static void x_input_free(void *data)
...
+
+   switch (id)
+   {
+#ifdef __linux__
+      case RETRO_SENSOR_ILLUMINANCE:
+         if (x11->illuminance_sensor)
+            return linux_get_illuminance_reading(x11->illuminance_sensor);
+#endif
+      default:
+         return 0.0f;
```
Comment (LibretroAdmin) [inline input/drivers/x11_input.c:497]: Same here.

**Review by LibretroAdmin — CHANGES_REQUESTED** [review]

File: input/drivers/linuxraw_input.c [hunk 11]
```diff
@@ -165,9 +166,58 @@ static void linuxraw_input_free(void *data)
...
+      case RETRO_SENSOR_ILLUMINANCE_ENABLE:
+         if (linuxraw->illuminance_sensor)
+            /* If the light sensor is already open, just set the rate */
+            linux_set_illuminance_sensor_rate(linuxraw->illuminance_sensor, rate);
+         else
+            linuxraw->illuminance_sensor = linux_open_illuminance_sensor(rate);
+
+         return linuxraw->illuminance_sensor != NULL;
+      default:
+         return false;
```
Comment (LibretroAdmin) [inline input/drivers/linuxraw_input.c:200]: This function should ALWAYS return something at the end. See previous comment about this.

File: input/drivers/linuxraw_input.c [hunk 12]
```diff
@@ -165,9 +166,58 @@ static void linuxraw_input_free(void *data)
...
+   if (!linuxraw)
+      return 0.0f;
+
+   switch (id)
+   {
+      case RETRO_SENSOR_ILLUMINANCE:
+         if (linuxraw->illuminance_sensor)
+            return linux_get_illuminance_reading(linuxraw->illuminance_sensor);
+      default:
+         return 0.0f;
```
Comment (LibretroAdmin) [inline input/drivers/linuxraw_input.c:217]: See previous comment

File: input/drivers/sdl_input.c [hunk 13]
```diff
@@ -336,9 +345,69 @@ static void sdl_input_free(void *data)
...
+            /* If we already have a sensor, just set the rate */
+            linux_set_illuminance_sensor_rate(sdl->illuminance_sensor, rate);
+         else
+            sdl->illuminance_sensor = linux_open_illuminance_sensor(rate);
+
+         return sdl->illuminance_sensor != NULL;
+#endif
+      /* Unsupported on non-Linux platforms */
+      default:
+         return false;
```
Comment (LibretroAdmin) [inline input/drivers/sdl_input.c:387]: See previous comment

File: input/drivers/sdl_input.c [hunk 14]
```diff
@@ -336,9 +345,69 @@ static void sdl_input_free(void *data)
...
+   switch (id)
+   {
+      case RETRO_SENSOR_ILLUMINANCE:
+#ifdef __linux__
+         if (sdl->illuminance_sensor)
+            return linux_get_illuminance_reading(sdl->illuminance_sensor);
+#endif
+      /* Unsupported on non-Linux platforms */
+      default:
+         return 0.0f;
```
Comment (LibretroAdmin) [inline input/drivers/sdl_input.c:407]: See previous comment

**Review by LibretroAdmin — CHANGES_REQUESTED** [review]

File: input/drivers/udev_input.c (hunk 7 above)
Comment (JesseTG) [reply input/drivers/udev_input.c:4051]: I'll happily do it, but is there a reason you prefer that style?

File: input/common/linux_common.c [hunk 15]
```diff
@@ -121,3 +156,222 @@ bool linux_terminal_disable_input(void)
...
+   sensor->millilux  = 0;
+   sensor->poll_rate = rate ? rate : DEFAULT_POLL_RATE;
+   sensor->thread    = NULL; /* We'll spawn a thread later, once we find a sensor */
+   sensor->done      = false;
+
+   devices = opendir(IIO_DEVICES_DIR);
+   if (!devices)
+   { /* If we couldn't find the IIO device directory... */
+      char errmesg[NAME_MAX_LENGTH];
+      strerror_r(errno, errmesg, sizeof(errmesg));
```
Comment (LibretroAdmin) [inline input/common/linux_common.c:214]: I don't see us ever using this before in the entire codebase. Why is this done and what is its purpose? Can you not just use strlcpy instead?

**Review by LibretroAdmin — CHANGES_REQUESTED** [review]

File: input/common/linux_common.c (hunk 15 above)
Comment (JesseTG) [reply input/common/linux_common.c:214]: I'm not sure I understand, "this" being what exactly? My use of `strerror_r`?

Comment (LibretroAdmin) [reply input/common/linux_common.c:214]: Yes, why is this necessary?

Comment (JesseTG) [reply input/common/linux_common.c:214]: [`opendir(3)`](https://man7.org/linux/man-pages/man3/opendir.3.html) sets `errno` upon failure, and [`strerror_r(3)`](https://man7.org/linux/man-pages/man3/strerror.3.html) is thread-safe.

Comment (LibretroAdmin) [reply input/common/linux_common.c:214]: Don't we have wrapper functions for the likes of opendir in libretro-common? I'd prefer we just go through that instead. retro_dirent IIRC.

Comment (JesseTG) [reply input/common/linux_common.c:214]: Ah, yes, I'd forgotten about those. I'll switch to them, not a problem.

**Pull request merged**